src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureConfig.kt

```diff
@@ -0,0 +1,143 @@
+    var returnType: Ty,
+    var visibility: RsVis? = null,
+    var isAsync: Boolean = false,
+    var isUnsafe: Boolean = false
+) {
+    val originalParameters = parameters.toList()
+    val parameters = parameters.toMutableList()
+
+    private val parametersText: String
+        get() = parameters.joinToString(", ") { "${it.patText}: ${it.type.renderInsertionSafe()}" }
+
+    fun signature(): String = buildString {
+        visibility?.let { append("${it.text} ") }
+
+        if (isAsync) {
+            append("async ")
+        }
+        if (isUnsafe) {
+            append("unsafe ")
+        }
+        append("fn $name$typeParametersText($parametersText)")
+        if (returnType !is TyUnit) {
+            append(" -> ${returnType.renderInsertionSafe()}")
+        }
+        append(whereClausesText)
+    }
+
+    private val typeParametersText: String
+        get() {
+            val typeParams = typeParameters()
+            if (typeParams.isEmpty()) return ""
+            return typeParams.joinToString(separator = ",", prefix = "<", postfix = ">") { it.text }
+        }
+
+    private val whereClausesText: String
+        get() {
+            val wherePredList = function.whereClause?.wherePredList ?: return ""
+            if (wherePredList.isEmpty()) return ""
+            val typeParams = typeParameters().map { it.declaredType }
+            if (typeParams.isEmpty()) return ""
+            val filtered = wherePredList.filter { it.typeReference?.type in typeParams }
+            if (filtered.isEmpty()) return ""
+            return filtered.joinToString(separator = ",", prefix = " where ") { it.text }
+        }
+
+    private fun typeParameters(): List<RsTypeParameter> {
+        val bounds = typeParameterBounds()
+        val paramAndReturnTypes = mutableSetOf<Ty>()
+        (parameters.map { it.type } + listOfNotNull(returnType)).forEach {
+            paramAndReturnTypes.addAll(it.types())
+            paramAndReturnTypes.addAll(it.dependTypes(bounds))
+        }
+        return function.typeParameters.filter { it.declaredType in paramAndReturnTypes }
+    }
+
+    private fun typeParameterBounds(): Map<Ty, Set<Ty>> =
+        function.typeParameters.associate { typeParameter ->
+            val type = typeParameter.declaredType
+            val bounds = mutableSetOf<Ty>()
+            typeParameter.bounds.flatMapTo(bounds) {
+                it.bound.traitRef?.path?.typeArguments?.flatMap { it.type.types() }.orEmpty()
+            }
+            type to bounds
+        }
+
+    fun createChangeInfo(): ChangeInfo = RsSignatureChangeInfo(this)
+
+    companion object {
+        fun create(function: RsFunction): RsFunctionSignatureConfig {
+            return RsFunctionSignatureConfig(function,
+                function.name ?: "",
+                function.valueParameters.mapIndexed { index, it ->
+                    Parameter(it.pat ?: RsPsiFactory(function.project).createPat("_"), it.typeReference?.type
```

Comment:
Currently, the whole function body is hard to read. Consider some reformatting, probably like that
```Kotlin
val psiFactory = RsPsiFactory(function.project)
val parameters = function.valueParameters.mapIndexed { index, parameter ->
    Parameter(parameter.pat ?: psiFactory.createPat("_"), parameter.typeReference?.type ?: TyUnknown, index)
}
return RsFunctionSignatureConfig(
    function,
    function.name.orEmpty(),
    parameters,
    function.returnType,
    function.vis,
    function.isAsync,
    function.isUnsafe
)
```